java/org/apache/catalina/filters/RateLimitFilter.java

```diff
@@ -0,0 +1,230 @@
+
+        param = config.getInitParameter(PARAM_BUCKET_REQUESTS);
+        if (param != null)
+            bucketRequests = Integer.parseInt(param);
+
+        param = config.getInitParameter(PARAM_ENFORCE);
+        if (param != null)
+            enforce = Boolean.parseBoolean(param);
+
+        param = config.getInitParameter(PARAM_STATUS_CODE);
+        if (param != null)
+            statusCode = Integer.parseInt(param);
+
+        param = config.getInitParameter(PARAM_STATUS_MESSAGE);
+        if (param != null)
+            statusMessage = param;
+
+        bucketCounter = new TimeBucketCounter(bucketDuration);
+
+        actualRequests = (int) Math.round(bucketCounter.getRatio() * bucketRequests);
+
+        log.info(sm.getString("rateLimitFilter.initialized",
+            super.getFilterName(), bucketRequests, bucketDuration, getActualRequests(),
+            getActualDurationInSeconds(), (!enforce ? "Not " : "") + "enforcing")
+        );
+    }
+
+    /**
+     * The <code>doFilter</code> method of the Filter is called by the container
+     * each time a request/response pair is passed through the chain due to a
+     * client request for a resource at the end of the chain. The FilterChain
+     * passed in to this method allows the Filter to pass on the request and
+     * response to the next entity in the chain.
+     * <p>
+     * A typical implementation of this method would follow the following
+     * pattern:- <br>
+     * 1. Examine the request<br>
+     * 2. Optionally wrap the request object with a custom implementation to
+     * filter content or headers for input filtering <br>
+     * 3. Optionally wrap the response object with a custom implementation to
+     * filter content or headers for output filtering <br>
+     * 4. a) <strong>Either</strong> invoke the next entity in the chain using
+     * the FilterChain object (<code>chain.doFilter()</code>), <br>
+     * 4. b) <strong>or</strong> not pass on the request/response pair to the
+     * next entity in the filter chain to block the request processing<br>
+     * 5. Directly set headers on the response after invocation of the next
+     * entity in the filter chain.
+     *
+     * @param request  The request to process
+     * @param response The response associated with the request
+     * @param chain    Provides access to the next filter in the chain for this
+     *                 filter to pass the request and response to for further
+     *                 processing
+     * @throws IOException      if an I/O error occurs during this filter's
+     *                          processing of the request
+     * @throws ServletException if the processing fails for any other reason
+     */
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+                            throws IOException, ServletException {
+
+        String ipAddr = request.getRemoteAddr();
```

Comment:
For tomcat, which behind a proxy (e.g. nginx), this filter is meaningless. I think we can write common method (check whether `x-forwarded-for`  header exists) that handle that case or allow user to specific a way that how to get really IP (specific header or others).

Reply:
Somewhat agree. I don't recall seeing a Connector attribute (or filter) that will provide the client remoteAddr. Which can also come i handy for CDN's which send the client IP via header too. It feels like that might need to be a new feature.

If x-forwarded-for is used, we need to be wary and document its ability to be insecure too. Such as bad actors sending this header randomized. In which case, documenting that we need to document the assumption that the header is value is trusted.

So by itself `String ipAddr = request.getRemoteAddr();` (I think) is OK ... if a way to implement request.getRemoteAddr() can be adjusted accordingly. 

The alternative is an option called ipHeader (horrible name, just a suggestion) which is the name of the header to use to get client IP address.

Reply:
Personally I was seeing this as a niche feature for Tomcat standalone. If there's a proxy, then it should be that component doing any rate limiting. Or did I miss something ?

Reply:
If expectation is standalone, then its just a documentation update to clarifying limitations. That should (optimistically)  eliminate security practitioners claiming new CVE's

Reply:
Tomcat already ships with an excellent RemoteIpFilter / RemoteIpValve and this filter does not attempt to replace them, but to work with them. The Filter documentation comes with a **WARNING** [1]:

> if Tomcat is behind a reverse proxy then you must
    make sure that the Rate Limit Filter sees the client IP address, so if for
    example you are using the Remote IP Filter, 
    then the filter mapping for the Rate Limit Filter must come <em>after</em>
    the mapping of the Remote IP Filter to ensure that each request has its IP
    address resolved before the Rate Limit Filter is applied

Did I get that wrong or am I correct that by having the RemoteIpFilter earlier in the Filter Chain then subsequent filters will see the "translated" address when calling `request.getRemoteAddr()`?  cc @rmaucher @ChristopherSchultz @markt-asf 

This Filter can work either in a standalone configuration, or behind a reverse proxy as long as the Remote Address is resolved using the methods mentioned above.  Even with reverse proxy it is useful to have this filter in place as it adds:

1. Configuration via web.xml which might be more convenient for those who are more familiar with Servlet containers than with web servers
2. A lenient mode in a way of `enforce=false`, allowing the user to configure the filter to only add a Request Attribute with the request count.  That allows the application to inspect the value and make a more intelligent decision since it knows more than the web server or the filter. 

[1] https://github.com/apache/tomcat/pull/607/files#diff-e79a16176f7d3d059a2bf4c43d5f50e9c107e27a2511379056ceea48b578c3e5R1001

Reply:
>  The Filter documentation comes with a WARNING [1]:

Sorry, I didn't look the doc carefully. 
> Did I get that wrong or am I correct that by having the RemoteIpFilter earlier in the Filter Chain then subsequent filters will see the "translated" address when calling request.getRemoteAddr()

It's ok to use with `RemoteIpFilter`.

BTW, I think we should also copy WARNING to Javadoc of `RateLimitFilter`. :)

Reply:
Great doc ! (of course, nobody reads it)